### PR TITLE
add socket way for module inter communicate

### PR DESCRIPTION
**Pull request by gy95**
Signed-off-by: gy95 <guoyao17@huawei.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Now modules communicate with each other only in the `channel` mode. So if module wants to communicate with another process, we must establish tcp/websocket connection itself

This PR will introduce `unix socket` mode. So different progresses could communicate using `unix socket`
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/github.com/kubeedge/beehive/pkg/core/socket/broker/broker_test.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	"net"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"github.com/kubeedge/beehive/pkg/core/model"
+	"src/github.com/kubeedge/beehive/pkg/core/socket/wrapper"
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/socket/broker/broker_test.go:14]:
```suggestion
	"github.com/kubeedge/beehive/pkg/core/socket/wrapper"
```

Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/socket/broker/broker_test.go:14]: ok，misspell

File: staging/src/github.com/kubeedge/beehive/pkg/common/type.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package common
+
+// define channel type
+const (
+	// MsgCtxTypeChannel message type channel
+	MsgCtxTypeChannel = "channel"
+	// MsgCtxTypeSocket message type socket
+	MsgCtxTypeSocket  = "tcp"
+	// MsgCtxTypeUS message type us
+	MsgCtxTypeUS = "unixpacket"
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/common/type.go:10]: These two types are all used now?

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_channel.go [hunk 3]
```diff
@@ -362,11 +372,16 @@ func (ctx *ChannelContext) addTypeChannel(module, group string, moduleCh chan mo
 }
 
 // AddModule adds module into module context
-func (ctx *ChannelContext) AddModule(module string) {
+func (ctx *ChannelContext) AddModule(module string, ch interface{}) {
```
[the remark this answers is not in the document]
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/context/context_channel.go:375]: for socket use, and in socket way, this parameter is connection, and this implement the interface
```
// AddModule add module
func (m *SocketModule) AddModule(module string, usConn interface{}) {
	if conn, ok := usConn.(wrapper.Conn); ok {
		m.store.Add(module, conn)
		return
	}
	klog.Errorf("failed to add module, bad us conn")
}
```
```
// ModuleContext is interface for context module management
type ModuleContext interface {
	AddModule(module string, ch interface{})
	AddModuleRemote(module string)
	AddModuleGroup(module, group string)
	Cleanup(module string)
}
```

File: staging/src/github.com/kubeedge/beehive/pkg/common/type.go (hunk 2 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/common/type.go:10]: yes

Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/common/type.go:10]: Oh, i'll keep the `"unixpacket"` and remove the `"tcp"`

Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/common/type.go]: done

File: staging/src/github.com/kubeedge/beehive/pkg/core/channel/context_channel.go [hunk 4]
```diff
@@ -45,6 +47,13 @@ func NewChannelContext() *ChannelContext {
 	}
 }
 
+func GetModuleChannel() *ChannelContext {
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/channel/context_channel.go:50]:
```suggestion
func GetChannelContext() *ChannelContext {
```

File: staging/src/github.com/kubeedge/beehive/pkg/core/channel/context_channel.go [hunk 5]
```diff
@@ -21,6 +21,9 @@ const (
 	TickerTimeoutDefault = 20 * time.Millisecond
 )
 
+var channelContext *ChannelContext
+var onceChannel sync.Once
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/channel/context_channel.go:25]:
```suggestion
var channelOnce sync.Once
```

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go [hunk 6]
```diff
@@ -18,27 +17,76 @@ type Module interface {
...
+	} else {
+		disabledModules[m.Name()] = info
+		klog.Infof("module %s is not register", m.Name())
 	}
 }
 
+// IsModuleEnabled return the module status
+func IsModuleEnabled(m Module) bool {
+	return m.Enable()
+}
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/module.go:77]: Remove this func, use `m.Enable()` directly.

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go [hunk 7]
```diff
@@ -18,27 +17,76 @@ type Module interface {
...
+// if not passed in parameter opts, default moduleType is "channel"
+func Register(m Module, opts ...string) {
+	var info moduleInfo
+	if len(opts) == 0 {
+		info = moduleInfo{
+			module:     m,
+			moduleType: common.MsgCtxTypeChannel,
+			remote:     false,
+		}
 	} else {
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/module.go:46]: The else here can be deleted.

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go [hunk 8]
```diff
@@ -18,27 +17,76 @@ type Module interface {
...
 	return modules
 }
+
+// Cleanup modules
+func Cleanup() {
+	return
+}
+
+// GetModule gets module
+func (m *moduleInfo) GetModule() Module {
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/module.go:90]: Who will call this func? If not, please delete

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go [hunk 9]
```diff
@@ -18,27 +17,76 @@ type Module interface {
...
 // GetModules gets modules map
-func GetModules() map[string]Module {
+func GetModules() map[string]moduleInfo {
 	return modules
 }
+
+// Cleanup modules
+func Cleanup() {
+	return
+}
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/module.go:87]: remove this func

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go [hunk 10]
```diff
@@ -18,27 +17,76 @@ type Module interface {
...
+			ModuleType: info.moduleType,
+			ModuleSocket: context.ModuleSocket{
+				IsRemote: info.remote,
+			},
+		}
+		coreContext.AddModule(&add)
+		//Assemble typeChannels for sendToGroup
+		coreContext.AddModuleGroup(m.Name(), info.module.Group())
+		modules[m.Name()] = info
+		klog.Infof("module %s registered", m.Name())
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/module.go:67]: Don't modify the log.

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go (hunk 8 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/module.go:90]: by metamanger_test.go, need to get the field `module`, which is not exported

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go (hunk 7 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/module.go:46]: need to distiguish moduleType by length.

File: staging/src/github.com/kubeedge/beehive/pkg/core/module.go (hunk 10 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/module.go]: all done except two nits

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go [hunk 11]
```diff
@@ -2,111 +2,334 @@ package context
...
+
+	ctx    gocontext.Context
+	cancel gocontext.CancelFunc
+}
 
 var (
 	// singleton
-	context *beehiveContext
-	once    sync.Once
+	beehiveContext *BeehiveContext
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:36]:
```suggestion
	globalContext *BeehiveContext
```

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go [hunk 12]
```diff
@@ -2,111 +2,334 @@ package context
...
+	"github.com/kubeedge/beehive/pkg/core/socket"
+	"github.com/kubeedge/beehive/pkg/core/socket/config"
 )
 
-// define channel type
-const (
-	MsgCtxTypeChannel = "channel"
-)
+// BeehiveContext is global context: only use for local cache to dispatch message
+type BeehiveContext struct {
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:19]:
```suggestion
type beehiveContext struct {
```
?

File: staging/src/github.com/kubeedge/beehive/pkg/core/core.go [hunk 13]
```diff
@@ -10,18 +10,35 @@ import (
...
+		}
+		context.AddModule(&module)
+		context.AddModuleGroup(info.module.Name(), info.module.Group())
+	}
+}
+
 // StartModules starts modules that are registered
 func StartModules() {
-	beehiveContext.InitContext(beehiveContext.MsgCtxTypeChannel)
+	coreContext := beehiveContext.GetGlobalContext()
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/core.go:36]: beehiveContext.InitContext([]string(beehiveContext.MsgCtxTypeChannel))

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go [hunk 14]
```diff
@@ -2,111 +2,334 @@ package context
...
-	context *beehiveContext
-	once    sync.Once
+	beehiveContext *BeehiveContext
+	once           sync.Once
 )
 
+// GetGlobalContext get global context instance
+func GetGlobalContext() *BeehiveContext {
+	return newContext()
+}
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:43]: Do we need this func?

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go [hunk 15]
```diff
@@ -2,111 +2,334 @@ package context
...
+
 // InitContext gets global context instance
 func InitContext(contextType string) {
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:46]:
```suggestion
func InitContext(contextTypes []string) {
```

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go (hunk 14 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:43]: yes, to get global variable `globalContext`

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go (hunk 15 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:46]: don't need slice format, once for one contextType.

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go (hunk 12 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:19]: this structure is also used by `staging/src/github.com/kubeedge/beehive/pkg/core/core.go`, need to be exported.

**Comment by gy95** [comment]
ping @fisherxu please review again 😄

File: staging/src/github.com/kubeedge/beehive/pkg/core/channel/context_channel.go [hunk 16]
```diff
@@ -32,8 +32,25 @@ type ChannelContext struct {
...
+	moduleChannels := make(map[string]map[string]chan model.Message)
+	anonChannels := make(map[string]chan model.Message)
+	channelContext = &ChannelContext{
+		channels:     channelMap,
+		typeChannels: moduleChannels,
+		anonChannels: anonChannels,
+	}
+}
+
+// tODO:
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/channel/context_channel.go:52]: Remove the codes above.

File: staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go [hunk 17]
```diff
@@ -2,111 +2,324 @@ package context
...
+func (ctx *BeehiveContext) GetContext(moduleType string, moduleName string) (ModuleContext, MessageContext) {
+	switch moduleType {
+	case common.MsgCtxTypeChannel:
+		context := channel.GetChannelContext()
+		return context, context
+	case common.MsgCtxTypeUS:
+		context := socket.NewSocketContext(moduleName)
+		return context, context
+	}
+	klog.Errorf("bad module type: %s", moduleType)
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/core/context/context_factory.go:323]:
```suggestion
	klog.Errorf("unsupported module type: %s", moduleType)
```

File: staging/src/github.com/kubeedge/beehive/go.mod [hunk 18]
```diff
@@ -3,9 +3,10 @@ module github.com/kubeedge/beehive
 go 1.14
 
 require (
+	github.com/go-chassis/go-archaius v1.5.1
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/go.mod:6]: Remove this repo.

File: staging/src/github.com/kubeedge/beehive/go.sum [hunk 19]
```diff
@@ -1,8 +1,78 @@
...
+github.com/gin-gonic/gin v1.4.0/go.mod h1:OW2EZn3DO8Ln9oIKOvM++LBO+5UPHJJDH72/q/3rZdM=
+github.com/go-chassis/cari v0.0.0-20201210041921-7b6fbef2df11/go.mod h1:MgtsEI0AM4Ush6Lyw27z9Gk4nQ/8GWTSXrFzupawWDM=
+github.com/go-chassis/foundation v0.2.2-0.20201210043510-9f6d3de40234/go.mod h1:2PjwqpVwYEVaAldl5A58a08viH8p27pNeYaiE3ZxOBA=
+github.com/go-chassis/foundation v0.2.2 h1:pu378+kf/NejqgHU9LogU6smWpjs+Mx0S99y27nW8OQ=
+github.com/go-chassis/foundation v0.2.2/go.mod h1:2PjwqpVwYEVaAldl5A58a08viH8p27pNeYaiE3ZxOBA=
+github.com/go-chassis/go-archaius v1.5.1 h1:1FrNyzzmD6o6BIjPF8uQ4Cc+u7qYIgQTpDk8uopBqfo=
+github.com/go-chassis/go-archaius v1.5.1/go.mod h1:QPwvvtBxvwiC48rmydoAqxopqOr93RCQ6syWsIkXPXQ=
+github.com/go-chassis/kie-client v0.0.0-20201210060018-938c7680a9ab/go.mod h1:UTdbtyN5ge/v9DmQzdVRxQP7z51Q4z6hyl+W6ZpUHFM=
+github.com/go-chassis/openlog v1.1.2 h1:LgGfwwOhpU8c6URV6ADpaRBPVY7Ph1C28jCQ6zzQawQ=
+github.com/go-chassis/openlog v1.1.2/go.mod h1:+eYCADVxWyJkwsFMUBrMxyQlNqW+UUsCxvR2LrYZUaA=
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/go.sum:20]: remove these repos and consider replace it with like `sigs.k8s.io/yaml`.

File: staging/src/github.com/kubeedge/beehive/pkg/common/config/config.go [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+//Event is a method get config value and logs it
+func (e EventListener) Event(event *event.Event) {
+	configValue := archaius.GetValue(event.Key)
+	for _, c := range ConfigChangeCallbacks {
+		c.Callback(event.Key, configValue)
+		klog.Infof("config value for %s", event.Key)
+	}
+}
```
Comment (fisherxu) [inline staging/src/github.com/kubeedge/beehive/pkg/common/config/config.go:188]: remove this files

File: staging/src/github.com/kubeedge/beehive/go.mod (hunk 18 above)
Comment (gy95) [reply staging/src/github.com/kubeedge/beehive/go.mod]: ok, please review again.

**Review by fisherxu — COMMENTED** [review]
/lgtm

**Pull request merged**